coroutines-extra/src/main/scala/org/coroutines/extra/AsyncAwait.scala

```diff
@@ -0,0 +1,104 @@
+package org.coroutines.extra
+
+
+
+import org.coroutines._
+import scala.annotation.unchecked.uncheckedVariance
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent._
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox.Context
+import scala.util.{ Success, Failure }
+
+
+
+object AsyncAwait {
+  /** Await the result of a future.
+   *
+   *  When called inside an `async` body, this function will block until its
+   *  associated future completes.
+   *
+   *  @return A coroutine that yields a tuple. `async` will assign this tuple's
+   *          second element to hold the completed result of the `Future` passed
+   *          into the coroutine. The coroutine will directly return the
+   *          result of the future.
+   */
+  def await[R]: Future[R] ~~> (Future[R], R) =
+    coroutine { (awaitedFuture: Future[R]) =>
+      yieldval(awaitedFuture)
+      var result: R = null.asInstanceOf[R]
+      awaitedFuture.value match {
+        case Some(Success(x)) => result = x
+        case Some(Failure(error)) => throw error
+        case None => sys.error("Future was not completed")
+      }
+      result
+    }
+
+  /** Calls `body`, blocking on any calls to `await`.
+   *
+   *  @param body  A coroutine to be invoked.
+   *  @return      A `Future` wrapping the result of the coroutine. The future fails
+   *               if `body` throws an exception or one of the `await`s takes a failed
+   *               future.
+   */
+  def asyncCall[Y, R](body: ~~~>[Future[Y], R]): Future[R] = {
+    val c = call(body())
+    val p = Promise[R]
+    def loop() {
+      if (!c.resume) {
+        c.tryResult match {
+          case Success(result) => p.success(result)
+          case Failure(exception) => p.failure(exception)
+        }
+      } else {
+        val awaitedFuture = c.value
+        if (awaitedFuture.isCompleted) {
+          awaitedFuture.value match {
+            case Some(Success(result)) =>
+              loop()
+            case Some(Failure(exception)) =>
```

Comment:
I think I'm misinterpreting the test you're describing because, as of 7fa77135d81c355bbee0121ada494bda88babb0f, this passes:

```scala
test("await should bubble up exceptions") {
  def thrower() = {
    throw new TestException
    Future(1)
  }

  var exceptionFound = false
  val future = async {
    try {
      await(thrower())
      ()
    } catch {
      case _: TestException => exceptionFound = true
    }
  }
  val r = Await.result(future, 1 seconds)
  assert(exceptionFound)
}
```